File: app/src/main/java/br/com/popularmovies/services/movieService/source/local/MovieLocalDataSource.java
package br.com.popularmovies.services.movieService.source.local;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import java.util.List;

import br.com.popularmovies.data.local.AppDatabase;
import br.com.popularmovies.data.model.ErrorResponse;
import br.com.popularmovies.data.model.OldResource;
import br.com.popularmovies.services.movieService.response.Movie;
import br.com.popularmovies.services.movieService.response.MovieReviews;
import br.com.popularmovies.services.movieService.response.MovieTrailers;
import br.com.popularmovies.services.movieService.response.Movies;
import br.com.popularmovies.services.movieService.source.MovieDataSource;
import br.com.popularmovies.utils.AppExecutors;

import static br.com.popularmovies.core.network.NetworkConstantsKt.GENERIC_ERROR_CODE;
import static br.com.popularmovies.movies.Constants.ROOM_MSG_ERROR;

public class MovieLocalDataSource implements MovieDataSource {
    private volatile static MovieLocalDataSource INSTANCE = null;
    private final MovieDao mMovieDao;

    private MovieLocalDataSource(Context context) {
        this.mMovieDao = AppDatabase.getInstance(context).movieDao();
    }

    public static MovieLocalDataSource getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (MovieLocalDataSource.class) {
                if (INSTANCE == null) {
                    INSTANCE = new MovieLocalDataSource(context);
                }
            }
        }
        return INSTANCE;
    }

    public static void destroyInstance() {
        INSTANCE = null;
    }

    @Override
    public LiveData<OldResource<Movies>> getMovies(final String orderBy) {
        final MediatorLiveData<OldResource<Movies>> movies = new MediatorLiveData<>();
        movies.postValue(OldResource.
                <Movies>loading());
        try {
            movies.addSource(mMovieDao.getMovies(), new Observer<List<Movie>>() {
                @Override
                public void onChanged(List<Movie> fetchedMovies) {
                    movies.postValue(OldResource.success(new Movies(fetchedMovies)));
                }
            });

        } catch (Exception e) {
            movies.postValue(OldResource.<Movies>error(new ErrorResponse(GENERIC_ERROR_CODE,
                    ROOM_MSG_ERROR)));
        }
        return movies;
    }

    @Override
    public LiveData<OldResource<Movie>> getMovie(final int movieId) {
        final MediatorLiveData<OldResource<Movie>> movie = new MediatorLiveData<>();
        movie.postValue(OldResource.
                <Movie>loading());
        try {
            movie.addSource(mMovieDao.getMovie(movieId), new Observer<Movie>() {
                @Override
                public void onChanged(Movie fetchedMovie) {
                    movie.postValue(OldResource.success(fetchedMovie));
                }
            });

        } catch (Exception e) {
            movie.postValue(OldResource.<Movie>error(new ErrorResponse(GENERIC_ERROR_CODE,
                    ROOM_MSG_ERROR)));
        }
        return movie;
    }

    @Override
    public LiveData<OldResource<MovieReviews>> getMovieReviews(int movieId) {
        return null;
    }

    @Override
    public LiveData<OldResource<Boolean>> saveToFavorites(final int movieId, final boolean status) {
        final MutableLiveData<OldResource<Boolean>> mMovie = new MutableLiveData<>();
        mMovie.postValue(OldResource.
                <Boolean>loading());
        try {
            AppExecutors.getInstance().diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    mMovieDao.saveFavorites(movieId, status);
                    mMovie.postValue(OldResource.
                            success(status));
                }
            });
        } catch (Exception e) {
            mMovie.postValue(OldResource.<Boolean>error(new ErrorResponse(GENERIC_ERROR_CODE,
                    ROOM_MSG_ERROR)));
        }
        return mMovie;
    }

    @Override
    public LiveData<OldResource<Void>> saveMovies(final List<Movie> movies) {
        final MutableLiveData<OldResource<Void>> mMovie = new MutableLiveData<>();
        mMovie.postValue(OldResource.
                <Void>loading());
        try {
            AppExecutors.getInstance().diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    mMovieDao.insertAllMovies(movies);
                    mMovie.postValue(OldResource.<Void>
                            success(null));
                }
            });
        } catch (Exception e) {
            mMovie.postValue(OldResource.<Void>error(new ErrorResponse(GENERIC_ERROR_CODE,
                    ROOM_MSG_ERROR)));
        }
        return mMovie;
    }

    @Override
    public LiveData<OldResource<Void>> saveMovie(final Movie movie) {
        final MutableLiveData<OldResource<Void>> mMovie = new MutableLiveData<>();
        mMovie.postValue(OldResource.
                <Void>loading());
        try {
            AppExecutors.getInstance().diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    mMovieDao.insertMovie(movie);
                    mMovie.postValue(OldResource.<Void>
                            success(null));
                }
            });
        } catch (Exception e) {
            mMovie.postValue(OldResource.<Void>error(new ErrorResponse(GENERIC_ERROR_CODE,
                    ROOM_MSG_ERROR)));
        }
        return mMovie;
    }

    @Override
    public LiveData<OldResource<Void>> removeMovie(final Movie movie) {
        final MutableLiveData<OldResource<Void>> mMovie = new MutableLiveData<>();
        mMovie.postValue(OldResource.
                <Void>loading());
        try {
            AppExecutors.getInstance().diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    mMovieDao.deleteMovie(movie);
                    mMovie.postValue(OldResource.<Void>
                            success(null));
                }
            });
        } catch (Exception e) {
            mMovie.postValue(OldResource.<Void>error(new ErrorResponse(GENERIC_ERROR_CODE,
                    ROOM_MSG_ERROR)));
        }
        return mMovie;
    }

    @Override
    public LiveData<OldResource<MovieTrailers>> getMovieTrailers(int movieId) {
        return null;
    }

}

File: app/src/main/java/br/com/popularmovies/services/movieService/source/local/MovieLocalDataSource.kt
package br.com.popularmovies.services.movieService.source.local

import br.com.popularmovies.core.network.local.AppDatabase
import br.com.popularmovies.datanetwork.config.GENERIC_ERROR_CODE
import br.com.popularmovies.datanetwork.config.GENERIC_MSG_ERROR_MESSAGE
import br.com.popularmovies.datanetwork.config.GENERIC_MSG_ERROR_TITLE
import br.com.popularmovies.datanetwork.models.base.AppError
import br.com.popularmovies.datanetwork.models.base.Result
import br.com.popularmovies.movies.Constants.ROOM_MSG_ERROR
import br.com.popularmovies.services.movieService.response.MovieTable
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MovieLocalDataSource @Inject constructor(appDatabase: AppDatabase) {
    private val mMovieDao: MovieDao = appDatabase.movieDao()

    private val error = AppError(
            codErro = GENERIC_ERROR_CODE,
            title = GENERIC_MSG_ERROR_TITLE,
            message = GENERIC_MSG_ERROR_MESSAGE
    )

    suspend fun getMovies(): Result<List<MovieTable>> {
        return try {
            Result.Success(mMovieDao.movies())
        } catch (exception: Exception) {
            Result.Error(error)
        }
    }

    suspend fun getFavoriteMovies(isFavorite: Boolean): Result<List<MovieTable>> {
        return try {
            Result.Success(mMovieDao.getFavoriteMovies(isFavorite))
        } catch (exception: Exception) {
            Result.Error(error)
        }
    }

    suspend fun getMovie(movieId: Int): Result<MovieTable> {
        return try {
            Result.Success(mMovieDao.getMovie(movieId))
        } catch (e: Exception) {
            Result.Error(AppError(GENERIC_ERROR_CODE, ROOM_MSG_ERROR))
        }
    }


    suspend fun saveToFavorites(movieTable: MovieTable): Result<Unit> {
        return try {
            val updatedRowsCount = mMovieDao.saveFavorites(movieTable.id, movieTable.isFavorite)
            Result.Success(updatedRowsCount)

        } catch (e: Exception) {
            Result.Error(AppError(GENERIC_ERROR_CODE, ROOM_MSG_ERROR))
        }
    }

    suspend fun insertMovie(movieTable: MovieTable): Result<Unit> {
        return try {
            Result.Success(mMovieDao.insertMovie(movieTable))
        } catch (e: Exception) {
            Result.Error(AppError(GENERIC_ERROR_CODE, ROOM_MSG_ERROR))
        }
    }

    suspend fun isMovieExists(movieId: Int): Result<Boolean> {
        return try {
            Result.Success(mMovieDao.isMovieExists(movieId))
        } catch (e: Exception) {
            Result.Error(AppError(GENERIC_ERROR_CODE, ROOM_MSG_ERROR))
        }
    }
}
